desc/coils.py

```diff
@@ -829,43 +853,74 @@ def flip(self, *args, **kwargs):
         """Compute magnetic field at a set of points.
 
         Parameters
         ----------
-        coords : array-like shape(n,3) or Grid
-            coordinates to evaluate field at [R,phi,Z] or [x,y,z]
+        grid : array-like shape(n,3) or Grid
+            Nodes to evaluate field at in [R,phi,Z] or [X,Y,Z] coordinates.
         params : dict or array-like of dict, optional
-            parameters to pass to curves, either the same for all curves,
-            or one for each member
+            Parameters to pass to coils, either the same for all coils or one for each.
         basis : {"rpz", "xyz"}
-            basis for input coordinates and returned magnetic field
-        grid : Grid, int or None or array-like, optional
-            Grid used to discretize coil, the same for all coils. If an integer, uses
-            that many equally spaced points.
+            Basis for input coordinates and returned magnetic field.
+        source_grid : Grid, int or None or array-like, optional
+            Grid used to discretize coils. If an integer, uses that many equally spaced
+            points. Should NOT include endpoint at 2pi.
 
         Returns
         -------
         field : ndarray, shape(n,3)
-            magnetic field at specified points, in either rpz or xyz coordinates
+            Magnetic field at specified nodes, in [R,phi,Z] or [X,Y,Z] coordinates.
+
         """
+        assert basis.lower() in ["rpz", "xyz"]
+        if hasattr(grid, "nodes"):
+            grid = grid.nodes
+        grid = jnp.atleast_2d(grid)
         if params is None:
             params = [get_params(["x_s", "x", "s", "ds"], coil) for coil in self]
             for par, coil in zip(params, self):
                 par["current"] = coil.current
-        if hasattr(coords, "nodes"):
-            coords = coords.nodes
-        coords = jnp.atleast_2d(coords)
 
-        def body(B, x):
-            B += self[0].compute_magnetic_field(
-                coords, params=x, basis=basis, grid=grid
+        # stellarator symmetry is easiest in [X,Y,Z] coordinates
+        if basis.lower() == "rpz":
+            grid_xyz = rpz2xyz(grid)
+        else:
+            grid_xyz = grid
+
+        # if stellarator symmetric, add reflected nodes from the other half field period
+        if self.sym:
+            normal = jnp.array(
+                [-jnp.sin(jnp.pi / self.NFP), jnp.cos(jnp.pi / self.NFP), 0]
+            )
+            grid_sym = (
+                grid_xyz @ reflection_matrix(normal).T @ reflection_matrix([0, 0, 1]).T
             )
-            return B, None
+            grid_xyz = jnp.vstack((grid_xyz, grid_sym))
 
-        B = scan(body, jnp.zeros(coords.shape), tree_stack(params))[0]
+        # field period rotation is easiest in [R,phi,Z] coordinates
+        grid_rpz = xyz2rpz(grid_xyz)
 
+        # sum the magnetic fields from each field period
+        B = jnp.zeros_like(grid_rpz)
+        for k in range(self.NFP):
```

Comment:
could also do a scan over NFP which will be faster to compile than a python loop

Reply:
Yeah I do something like that for the singular integral stuff: https://github.com/PlasmaControl/DESC/blob/b245323b4c177f20f0becea1c055536dfeaf43d8/desc/singularities.py#L364-L369

You'd have the regular base grid and just shift it by i*2pi/NFP where i is the loop step